### PR TITLE
pki: When a role sets key_type to any ignore key_bits value when signing a csr

### PR DESCRIPTION
**Pull request by stevendpclark**
This addresses issue #16237. 

When signing a CSR using /sign/<:role name> api method we validate that the CSR matches up with the role's `key_type` and `key_bits` values. If the `key_type` is set to `any`, these checks don't really make sense, and these checks were effectively bypassed when the `key_bits` value was set to current default value of 0.

The issue that was missed was the default value for created role's `key_bits` parameter was changed in 1.10 from 2048 to 0. So effectively the fix we previously made within [PR#14875](https://github.com/hashicorp/vault/pull/14875) addressed the issue, but only when the role was created in Vault 1.10 and higher. 

Now we bypass the validation for the role's `key_bits` value when signing CSRs if the `key_type` is set to `any`. We still validate the key is at least 2048 for RSA backed CSRs as we did in 1.9.x and lower.

### CODE REVIEW

**Review by kitography — COMMENTED** [review]
Confused?

File: builtin/logical/pki/cert_util.go [hunk 1]
```diff
@@ -836,22 +836,22 @@ func signCert(b *backend,
...
+			actualKeyType, 0, data.role.SignatureBits); err != nil {
 			return nil, errutil.InternalError{Err: fmt.Sprintf("unknown internal error updating default values: %v", err)}
 		}
 
 		// We're using the KeyBits field as a minimum value, and P-224 is safe
 		// and a previously allowed value. However, the above call defaults
 		// to P-256 as that's a saner default than P-224 (w.r.t. generation).
 		// So, override our fake Role value if it was previously zero.
-		if actualKeyType == "ec" && roleBitsWasZero {
+		if actualKeyType == "ec" {
```
[the remark this answers is not in the document]
Comment (stevendpclark) [reply builtin/logical/pki/cert_util.go:854]: Agreed all this code is a tad confusing. The api docs specify the following

> signing existing CSRs, any can be specified to allow keys of either type and with any bit size (subject to >1024 bits for RSA keys).

So we are within the `any` block so we are resetting the minimum size of `key_bits` no matter what the role had as a value to the lowest values we feel comfortable supporting for EC keys like we do for RSA keys. 

I'll tweak the comment a bit dropping the "if it was previously zero" that I missed.

[the remark this answers is not in the document]
Comment (kitography) [reply builtin/logical/pki/cert_util.go:856]: Slack conversation: 

[Kit Haines](https://app.slack.com/team/U02MQNJ235W)
  [4:46 PM](https://hashicorp.slack.com/archives/D02NU3K2K1T/p1657226804029859)
I’m no longer sure if we are referring to “any” as keyBits or keyType
[4:47](https://hashicorp.slack.com/archives/D02NU3K2K1T/p1657226836106419)
if we have set keyType “any” and keyBits “256" would we want to update that on the role?
[4:47](https://hashicorp.slack.com/archives/D02NU3K2K1T/p1657226842668459)
or is that role not saved?


[Steve Clark](https://app.slack.com/team/U02FBT0L3SR)
  [4:47 PM](https://hashicorp.slack.com/archives/D02NU3K2K1T/p1657226856627989)
ha, yeah okay. When keyType == ‘any’ basically we just enforce mins for key bits no matter what the role says
[4:48](https://hashicorp.slack.com/archives/D02NU3K2K1T/p1657226881632229)
That role that we update in the code you are seeing it thrown away
[4:49](https://hashicorp.slack.com/archives/D02NU3K2K1T/p1657226946795899)
Alex had a different fix which might be clearer, I’m not sure. He was going to change the getRole function to always return a value of 0 for KeyBits if the role had a KeyType == ‘any’


[Kit Haines](https://app.slack.com/team/U02MQNJ235W)
  [4:50 PM](https://hashicorp.slack.com/archives/D02NU3K2K1T/p1657227017357989)
^ I think I’d fine that clearer, but given we are throwing away the changes, I guess it doesn’t actually matter

**Review by kitography — APPROVED** [review]
LGTM

File: website/content/api-docs/secret/pki.mdx [hunk 2]
```diff
@@ -2338,7 +2338,7 @@ request is denied.
   generated private keys and the type of key expected for submitted CSRs.
   Currently, `rsa`, `ec`, and `ed25519` are supported, or when signing
   existing CSRs, `any` can be specified to allow keys of either type
-  and with any bit size (subject to >1024 bits for RSA keys).
+  and with any bit size (subject to >1024 bits for RSA keys or >= 224 for EC keys).
```
Comment (cipherboy) [inline website/content/api-docs/secret/pki.mdx:2341]: Isn't this 2048 for RSA keys? I think we have a test that errs for 1024:

```
backend_test.go:555:				"csr":         csrPem1024,
backend_test.go:564:					return fmt.Errorf("received an error but not about a 1024-bit key, error was: %s", resp.Data["error"].(string))
```

File: changelog/16246.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+```release-note:bug
+secret/pki: Do not fail validation with a legacy key_bits default value when signing CSRs
```
Comment (cipherboy) [inline changelog/16246.txt:2]: Should probably clarify this is only for roles with `any` keytype?

File: builtin/logical/pki/cert_util.go [hunk 4]
```diff
@@ -836,22 +836,24 @@ func signCert(b *backend,
...
+		// meaningful value.
 		//
-		// This is mandatory because on some roles, with KeyType any, we'll
-		// set a default SignatureBits to 0, but this will need to be updated
-		// in order to behave correctly during signing.
-		roleBitsWasZero := data.role.KeyBits == 0
-		if data.role.KeyBits, data.role.SignatureBits, err = certutil.ValidateDefaultOrValueKeyTypeSignatureLength(actualKeyType, data.role.KeyBits, data.role.SignatureBits); err != nil {
+		// We ignore the role's original KeyBits value if the KeyType is any
+		// as legacy (pre-1.10) roles had default values that made sense only
+		// for RSA keys (2048) and the older code paths ignored the role value
```
Comment (cipherboy) [inline builtin/logical/pki/cert_util.go:843]: Might want to clarify that this is key_bits? My 2c. 

```suggestion
		// for RSA keys (key_bits=2048) and the older code paths ignored the role value
```

File: changelog/16246.txt (hunk 3 above)
Comment (stevendpclark) [reply changelog/16246.txt]: Done.

**Review by cipherboy — APPROVED** [review]

**Pull request merged**